All pravega-segmentstores are showing IllegalArgumentException frequently after starting IO client

Observing **IllegalArgumentException** execption frequently in all segmentstores(ERROR i.p.s.server.writer.StorageWriter - StorageWriter[28]: Iteration[9873].Error.) after starting IO client

It looks like after starting IO client all the segmentstores start throwing below exceptions and due to this after some time all readers are move to dead state

Deployment Details:-

```
bookkeeper:0.5.0-2222.8ba5431
pravega:0.5.0-2222.8ba5431
zookeeper:3.5.4-beta-operator
pravega-operator:0.3.2
zookeeper-operator:0.2.1
```

Environment details: PKS / K8 with medium cluster:

```
3 master: xlarge: 4 CPU, 16 GB Ram, 32 GB Disk
5 worker: 2xlarge: 8 CPU, 32 GB Ram, 64 GB Disk
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
Segmentstore Exception log, For complete log please find the attachment
```
2019-04-19 13:03:26,584 1698786 [storage-io-104] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[28]: Iteration[9875].Error.
java.lang.IllegalArgumentException: Offset 213617384 + length 34960256 is beyond the last offset 213617384 of the segment.
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:458)
        at io.pravega.segmentstore.storage.rolling.RollingStorage.read(RollingStorage.java:151)
        at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$read$8(AsyncStorageWrapper.java:128)
        at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$execute$13(AsyncStorageWrapper.java:184)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
2019-04-19 13:03:28,098 1700300 [core-20] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[28]: Iteration[9876].Error.
java.lang.IllegalArgumentException: Offset 213617384 + length 34960256 is beyond the last offset 213617384 of the segment.
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:458)
        at io.pravega.segmentstore.storage.rolling.RollingStorage.read(RollingStorage.java:151)
        at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$read$8(AsyncStorageWrapper.java:128)
        at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$execute$13(AsyncStorageWrapper.java:184)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
2019-04-19 13:03:28,212 1700414 [core-1] INFO  i.p.s.server.logs.DurableLog - DurableLog[28]: MetadataCheckpointOperation durably stored.

2019-04-19 13:12:57,420 2269622 [epollEventLoopGroup-11-3] INFO  i.p.s.s.h.handler.AppendProcessor - Received hello from connection: [id: 0xf5f8bafe, L:/172.24.75.15:12345 - R:/172.24.114.3:56706]
2019-04-19 13:12:57,420 2269622 [epollEventLoopGroup-11-3] INFO  i.p.s.s.h.handler.AppendProcessor - Setting up appends for writer: 29c1b333-1cad-44c2-b8f2-a772de3ef543 on segment: longevity/smallKey/0.#epoch.0
2019-04-19 13:12:57,423 2269625 [epollEventLoopGroup-11-3] ERROR i.p.s.s.h.h.ServerConnectionInboundHandler - Caught exception on connection:
io.netty.handler.codec.DecoderException: java.lang.NullPointerException
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:472)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:426)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
        at io.pravega.shared.protocol.netty.ExceptionLoggingHandler.channelRead(ExceptionLoggingHandler.java:37)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:799)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:421)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:321)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException: null
        at io.pravega.shared.protocol.netty.WireCommandType.readFrom(WireCommandType.java:122)
        at io.pravega.shared.protocol.netty.CommandDecoder.parseCommand(CommandDecoder.java:55)
        at io.pravega.shared.protocol.netty.CommandDecoder.decode(CommandDecoder.java:36)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
        ... 26 common frames omitted
2019-04-19 13:12:57,424 2269626 [epollEventLoopGroup-11-3] ERROR i.p.s.s.h.h.ServerConnectionInboundHandler - Caught exception on connection:
io.netty.handler.codec.DecoderException: io.pravega.shared.protocol.netty.InvalidMessageException: Unknown wire command: -1393754107
```

--- a/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/rolling/RollingStorage.java
+++ b/segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/rolling/RollingStorage.java
@@ -139,10 +139,11 @@ public class RollingStorage implements SyncStorage {
                     offset, bufferOffset, length, buffer.length));
         }
 
-        if (h.isReadOnly() && !h.isSealed() && offset + length > h.length()) {
-            // We have a non-sealed read-only handle. It's possible that the SegmentChunks may have been modified since
-            // the last time we refreshed it, and we received a request for a read beyond our last known offset. Reload
-            // the handle before attempting the read.
+        if (!h.isSealed() && offset + length > h.length()) {
+            // We have a non-sealed handle (read-only or read-write). It's possible that the SegmentChunks may have been
+            // modified since the last time we refreshed it, and we received a request for a read beyond our last known offset.
+            // This could happen if the Segment was modified using a different handle or a previous write did succeed but was
+            // reported as having failed. Reload the handle before attempting the read so that we have the most up-to-date info.
             val newHandle = (RollingSegmentHandle) openRead(handle.getSegmentName());
             h.refresh(newHandle);
             log.debug("Handle refreshed: {}.", h);
@@ -250,7 +251,7 @@ public class RollingStorage implements SyncStorage {
         // If the header file already exists, then it's OK if it's empty (probably a remnant from a previously failed
         // attempt); in that case we ignore it and let the creation proceed.
         SegmentHandle headerHandle = null;
-        RollingSegmentHandle retValue  = null;
+        RollingSegmentHandle retValue;
         try {
             try {
                 headerHandle = this.baseStorage.create(headerName);

--- a/segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/rolling/RollingStorageTests.java
+++ b/segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/rolling/RollingStorageTests.java
@@ -90,20 +90,40 @@ public class RollingStorageTests extends RollingStorageTestBase {
         @Cleanup
         val s = new RollingStorage(baseStorage, DEFAULT_ROLLING_POLICY);
         s.initialize(1);
-        s.create(SEGMENT_NAME);
-        val h1 = s.openWrite(SEGMENT_NAME);
-        val h2 = s.openWrite(SEGMENT_NAME); // Open now, before writing, so we force a refresh.
 
+        // We use this handle for writing.
+        val wh1 = s.create(SEGMENT_NAME);
+
+        // We use these handles for attempting to write in parallel or read. Open them now, before writing, so we force refresh.
+        val wh2 = s.openWrite(SEGMENT_NAME); // We use this to write in parallel.
+        val wh3 = s.openWrite(SEGMENT_NAME); // We use this to read using a Write Handle.
+        val rh1 = s.openRead(SEGMENT_NAME); // We use this to read using a Read Handle.
+
+        // Write data.
         byte[] data = "data".getBytes();
-        s.write(h1, 0, new ByteArrayInputStream(data), data.length);
-        s.write(h2, data.length, new ByteArrayInputStream(data), data.length);
+        s.write(wh1, 0, new ByteArrayInputStream(data), data.length);
+        s.write(wh2, data.length, new ByteArrayInputStream(data), data.length);
 
         // Check that no file has exceeded its maximum length.
         byte[] expectedData = new byte[data.length * 2];
         System.arraycopy(data, 0, expectedData, 0, data.length);
         System.arraycopy(data, 0, expectedData, data.length, data.length);
 
-        checkWrittenData(expectedData, h2, s);
+        // Read using the handle we just used for writing.
+        checkWrittenData(expectedData, wh2, s);
+
+        // Read using an out-of-date read handle.
+        checkWrittenData(expectedData, rh1, s);
+
+        // Read using an out-of-date write handle.
+        checkWrittenData(expectedData, wh3, s);
+
+        // And then verify we can still use that write handle for additional writing.
+        s.write(wh3, expectedData.length, new ByteArrayInputStream(data), data.length);
+        byte[] finalExpectedData = new byte[expectedData.length + data.length];
+        System.arraycopy(expectedData, 0, finalExpectedData, 0, expectedData.length);
+        System.arraycopy(data, 0, finalExpectedData, expectedData.length, data.length);
+        checkWrittenData(finalExpectedData, wh3, s);
     }
 
     /**

